lib/auth/auth.go

```diff
@@ -1214,6 +1225,92 @@ func (a *Server) DeleteMFADeviceWithToken(ctx context.Context, req *proto.Delete
 	return nil
 }
 
+// AddMFADeviceWithToken adds a mfa device for the user defined in token.
+func (a *Server) AddMFADeviceWithToken(ctx context.Context, req *proto.AddMFADeviceWithTokenRequest) error {
+	token, err := a.GetUserToken(ctx, req.GetTokenID())
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Restrict to certain kinds of user token.
+	if token.GetSubKind() != UserTokenTypePrivilege {
+		return trace.BadParameter("invalid token")
+	}
+
+	if token.Expiry().Before(a.clock.Now().UTC()) {
+		return trace.BadParameter("expired token")
+	}
+
+	authPref, err := a.GetAuthPreference(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	secondFactor := authPref.GetSecondFactor()
+	var newDevice *types.MFADevice
+
+	switch req.GetSecondFactorCred().(type) {
+	case *proto.AddMFADeviceWithTokenRequest_U2FRegisterResponse:
+		if secondFactor == constants.SecondFactorOTP {
```

Comment:
Is this if statement correct? I'm confused by the error that says "cluster only allows OTP" because the second factor in this case is OTP. Same with U2F below.

I think the bad parameter error should say "cluster only allows U2F for second factor" - right?

Reply:
It's strange for an user to reply to an U2F challenge the cluster didn't supply - they shouldn't have the challenge and it'll never pass anyway.

I guess this could happen for OTP, so maybe let's handle only that. As stated in other comments, though, I think a single method should handle this (it used be addMFADeviceRegisterChallenge).